syntax = "proto3";
import "automanlib_wrappers.proto";

/*****
This .proto file defins the rpc service and the messages sent
and received by the various rpc method calls. 
*****/


/*****
This service is a prototype of the Estimation functionality of
AutoMan only. 
RPC methods:

KillServer 
	This method shuts the rpc server down
parameters:
	None
returns:
	ServerStatusResponse
		Indicates the current status of the server, "KILLED" if the
		method was successful

RegisterAdapter 
	Registers a crowdsource back-end for the AutoMan server to use for
	authentication
parameters:
	AdapterCredentials 
returns:
	RegistrationResponse
		Indicates if the credentials were registered successfully or not


ServerStatus 
	Reports the status of the server
parameters:
	None 
returns:
	ServerStatusResponse
		Indicates the current status of the server, either "RUNNING" or 
		"KILLED"


SubmitTask 
	Submits a task to the AutoMan server to post to the crowdsource back-end
parameters:
	AutomanTask 
returns:
	TaskResponse
		A response from the Automan server on the submitted task
		If the response code is VALID, the 
		task was completed successfully and one of task_outcome will be set depending
		on the respective task.
		If the response code was ERROR, an error occured and the err_code and err_msg
		fields will be set.
		If the response code was EXCEPTION, an exception occured and the excep_code
		and excep_msg fields will be set.

*****/
service EstimationPrototype {
	rpc KillServer(Empty) returns (ServerStatusResponse) {}
	rpc RegisterAdapter(AdapterCredentials) returns (RegistrationResponse) {}
	rpc SubmitTask(AutomanTask) returns (TaskResponse) {}
	rpc ServerStatus(Empty) returns (ServerStatusResponse) {}
}


/*
This message class is a wrapper for the different types of 
tasks that can be submitted 
*/
message AutomanTask{
	oneof task_type{
		EstimateTask est_task = 1;
		MultiestimateTask mest_task = 2;
		FreetextTask free_task = 3;
		FreetextDistTask free_dist_task = 4;
		RadioTask radio_task = 5;
		RadioDistTask radio_dist_task = 6;
		CheckboxTask check_task = 7;
		CheckboxDistTask check_dist_task = 8;
	}
}


/*
This message class is the server response to a submitted task
fields:
return_code 				The type of response from the server

*_outcome 					* is replaced by est, mest, free, etc. from the oneof field. The outcome
							of the AutoMan task, set if return_code is VALID

err_msg 					A message about the type of error that occured, set if return_code is 
							ERROR

err_code 					A code representing the type of error that occured, set if return_code is 
							ERROR

excep_msg 					A message about the type of exception that occured, set if return_code is 
							EXCEPTION

excep_code 					A code representing the type of exception that occured, set if return_code is 
							EXCEPTION
*/
message TaskResponse{
	enum TaskReturnCode{
		UNDEFINED_RESP_CODE = 0;
		VALID = 1;
		ERROR = 2;
		EXCEPTION = 3;
	} 
	enum ExceptionCode{
		UNDEFINED_EXCPTION_CODE = 0;
	}
	enum ErrorCode{
		UNDEFINED_ERROR_CODE = 0;
		NO_CREDENTIALS_REGISTERED = 1;
	}

	TaskReturnCode return_code = 1;
	oneof task_outcome{
		EstimateOutcome est_outcome = 2;
		MultiestimateOutcome mest_outcome = 3;
		RadioOutcome rad_outcome = 4;
		RadioDistOutcome raddist_outcome = 5;
		FreetextOutcome free_outcome = 6;
		FreetextDistOutcome freedist_outocme = 7;
		CheckboxOutcome chec_outcome = 8;
		CheckboxDistOutcome checdist_outcome = 9;
	}

	ExceptionCode excep_code = 16;
	ErrorCode err_code = 17;
	string err_msg = 18;
	string excep_msg = 19;
}


/*
This message class is the server response to registering an adapter
fields:
return_code 				Indicates whether the adapter was registered successfully
							or not
message 					error message, set if the adapter could not be added
*/
message RegistrationResponse{
	enum RegReturnCode{
		UNDEFINED_CODE = 0;
		OKAY = 1;
		FAILED = 2;
	} 
	RegReturnCode return_code = 1;
 	string err_msg = 2;
}


This message class is the server response to registering an adapter
fields:
return_code 				Indicates the status of the server, either
							RUNNING or KILLED depending of the server process
							is in a alive or has been shutdown
*/
message ServerStatusResponse{
	enum StatReturnCode{
		UNDEFINED_STATUS_CODE = 0;
		RUNNING = 1;
		KILLED = 2;
	} 
	StatReturnCode return_code = 1;
}

message Empty{
	
}